plugins/state_history_plugin/include/eosio/state_history_plugin/state_history_log.hpp

```diff
@@ -87,39 +87,47 @@ class state_history_log {
+                  "missed a block in ${name}.log", ("name", name));
 
-      if (_begin_block != _end_block && block_num > _begin_block) {
-         if (block_num == _end_block) {
-            EOS_ASSERT(prev_id == last_block_id, chain::plugin_exception, "missed a fork change in ${name}.log",
-                       ("name", name));
-         } else {
-            state_history_log_header prev;
-            get_entry(block_num - 1, prev);
-            EOS_ASSERT(prev_id == prev.block_id, chain::plugin_exception, "missed a fork change in ${name}.log",
-                       ("name", name));
+         if (_begin_block != _end_block && block_num > _begin_block) {
+            if (block_num == _end_block) {
+               EOS_ASSERT(prev_id == last_block_id, chain::plugin_exception, "missed a fork change in ${name}.log",
+                        ("name", name));
+            } else {
+               state_history_log_header prev;
+               get_entry(block_num - 1, prev);
+               EOS_ASSERT(prev_id == prev.block_id, chain::plugin_exception, "missed a fork change in ${name}.log",
+                        ("name", name));
+            }
          }
-      }
 
-      if (block_num < _end_block)
-         truncate(block_num);
-      log.seek_end(0);
-      uint64_t pos = log.tellp();
-      write_header(header);
-      write_payload(log);
-      uint64_t end = log.tellp();
-      EOS_ASSERT(end == pos + state_history_log_header_serial_size + header.payload_size, chain::plugin_exception,
-                 "wrote payload with incorrect size to ${name}.log", ("name", name));
-      log.write((char*)&pos, sizeof(pos));
+         if (block_num < _end_block)
+            truncate(block_num);
+         log.seek_end(0);
+         uint64_t pos = log.tellp();
+         write_header(header);
+         write_payload(log);
+         uint64_t end = log.tellp();
+         EOS_ASSERT(end == pos + state_history_log_header_serial_size + header.payload_size, chain::plugin_exception,
+                  "wrote payload with incorrect size to ${name}.log", ("name", name));
+         log.write((char*)&pos, sizeof(pos));
 
-      index.seek_end(0);
-      index.write((char*)&pos, sizeof(pos));
-      if (_begin_block == _end_block)
-         _begin_block = block_num;
-      _end_block    = block_num + 1;
-      last_block_id = header.block_id;
+         index.seek_end(0);
+         index.write((char*)&pos, sizeof(pos));
+         if (_begin_block == _end_block)
+            _begin_block = block_num;
+         _end_block    = block_num + 1;
+         last_block_id = header.block_id;
+      }
+      catch(const chain::plugin_exception& e) {
+         elog( "chain::plugin_exception: ${details}", ("details", e.to_detail_string()) );
+         elog("State history encountered an Error which it cannot recover from.  Please resolve the error and relaunch "
+              "the process");
+         appbase::app().quit();
```

Comment:
I think you should throw like you had before but also call `app().quit()`. You don't want the block produced because on restart you need to reprocess the block so it can be written to SHiP.